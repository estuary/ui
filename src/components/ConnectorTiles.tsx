import { AddBox, Cable, OpenInNew } from '@mui/icons-material';
import {
    Box,
    Button,
    Grid,
    Paper,
    Skeleton,
    Stack,
    SxProps,
    Theme,
    Typography,
    useMediaQuery,
    useTheme,
} from '@mui/material';
import ConnectorToolbar from 'components/ConnectorToolbar';
import useEntityCreateNavigate from 'components/shared/Entity/useEntityCreateNavigate';
import {
    darkGlassBkgColor,
    darkGlassBkgColorIntensified,
    slate,
} from 'context/Theme';
import { useQuery, useSelect } from 'hooks/supabase-swr';
import {
    ConnectorWithTagDetailQuery,
    CONNECTOR_WITH_TAG_QUERY,
} from 'hooks/useConnectorWithTagDetail';
import { useEffect, useMemo, useRef, useState } from 'react';
import { FormattedDate, FormattedMessage, useIntl } from 'react-intl';
import { CONNECTOR_NAME, defaultTableFilter, TABLES } from 'services/supabase';
import {
    BaseComponentProps,
    ENTITY_WITH_CREATE,
    SortDirection,
    TableIntlConfig,
    TableState,
    TableStatuses,
} from 'types';
import { hasLength } from 'utils/misc-utils';
import { getEmptyTableHeader, getEmptyTableMessage } from 'utils/table-utils';

interface ConnectorTilesProps {
    cardWidth: number;
    cardsPerRow: number;
    gridSpacing: number;
    protocolPreset?: ENTITY_WITH_CREATE;
    replaceOnNavigate?: boolean;
}

type TileProps = BaseComponentProps;

const intlConfig: TableIntlConfig = {
    header: 'connectors.main.message1',
    message: 'connectors.main.message2',
};

const imageBackgroundSx: SxProps<Theme> = {
    width: '100%',
    height: 125,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: 2,
    borderRadius: 5,
    background: (theme) =>
        theme.palette.mode === 'dark'
            ? 'rgba(172, 199, 220, 0.30)' // Brighter than desired to improve MySQL visibility.
            : slate[25],
};

function Tile({ children }: TileProps) {
    return (
        <Paper
            elevation={0}
            sx={{
                'height': '100%',
                'borderRadius': 5,
                'background': (theme) =>
                    theme.palette.mode === 'dark'
                        ? darkGlassBkgColor
                        : slate[50],
                'padding': 1,
                '&:hover': {
                    background: (theme) =>
                        theme.palette.mode === 'dark'
                            ? darkGlassBkgColorIntensified
                            : 'rgba(172, 199, 220, 0.45)',
                },
            }}
        >
            <Stack
                style={{
                    height: '100%',
                    justifyContent: 'space-between',
                }}
            >
                {children}
            </Stack>
        </Paper>
    );
}

function ConnectorTiles({
    cardWidth,
    cardsPerRow,
    gridSpacing,
    protocolPreset,
    replaceOnNavigate,
}: ConnectorTilesProps) {
    const navigateToCreate = useEntityCreateNavigate();
    const isFiltering = useRef(false);
    const intl = useIntl();

    const theme = useTheme();
    const belowMd = useMediaQuery(theme.breakpoints.down('md'));

    const [protocol, setProtocol] = useState<string | null>(
        protocolPreset ?? null
    );
    const [searchQuery, setSearchQuery] = useState<string | null>(null);
    const [sortDirection, setSortDirection] = useState<SortDirection>('asc');
    const [columnToSort, setColumnToSort] =
        useState<keyof ConnectorWithTagDetailQuery>(CONNECTOR_NAME);

    const [tableState, setTableState] = useState<TableState>({
        status: TableStatuses.LOADING,
    });

    const primaryCtaClick = (row: ConnectorWithTagDetailQuery) => {
        navigateToCreate(
            row.connector_tags[0].protocol,
            row.connector_tags[0].id,
            replaceOnNavigate
        );
    };

    const liveSpecQuery = useQuery<ConnectorWithTagDetailQuery>(
        TABLES.CONNECTORS,
        {
            columns: CONNECTOR_WITH_TAG_QUERY,
            filter: (query) => {
                return defaultTableFilter<ConnectorWithTagDetailQuery>(
                    query,
                    [columnToSort],
                    searchQuery,
                    columnToSort,
                    sortDirection,
                    undefined,
                    { column: 'connector_tags.protocol', value: protocol }
                );
            },
        },
        [searchQuery, columnToSort, sortDirection, protocol]
    );

    const { data: useSelectResponse, isValidating } = useSelect(liveSpecQuery);
    const selectData = useMemo(
        () => (useSelectResponse ? useSelectResponse.data : []),
        [useSelectResponse]
    );

    useEffect(() => {
        if (selectData.length > 0) {
            setTableState({ status: TableStatuses.DATA_FETCHED });
        } else if (isFiltering.current) {
            setTableState({ status: TableStatuses.UNMATCHED_FILTER });
        } else {
            setTableState({ status: TableStatuses.NO_EXISTING_DATA });
        }
    }, [selectData, isValidating]);

    const gridContainerWidth = belowMd
        ? '100%'
        : cardWidth * cardsPerRow + 8 * gridSpacing * (cardsPerRow + 1);

    const skeletonTileCount = useMemo(
        () => 12 / (belowMd ? 6 : 12 / cardsPerRow),
        [belowMd, cardsPerRow]
    );

    return (
        <Grid
            container
            spacing={gridSpacing}
            paddingRight={2}
            width={gridContainerWidth}
            margin="auto"
        >
            <Grid item xs={12}>
                <ConnectorToolbar
                    belowMd={belowMd}
                    gridSpacing={gridSpacing}
                    setColumnToSort={setColumnToSort}
                    hideProtocol={!!protocolPreset}
                    setProtocol={setProtocol}
                    setSortDirection={setSortDirection}
                    setSearchQuery={setSearchQuery}
                />
            </Grid>

            {hasLength(selectData) ? (
                selectData
                    .map((row, index) => (
                        <Grid
                            key={`connector-tile-${index}`}
                            item
                            xs={6}
                            md={12 / cardsPerRow}
                        >
                            <Tile>
                                <Box sx={imageBackgroundSx}>
                                    {row.image ? (
                                        <img
                                            src={row.image}
                                            loading="lazy"
                                            alt=""
                                            style={{
                                                width: 'auto',
                                                maxHeight: 75,
                                                padding: '0 1rem',
                                            }}
                                        />
                                    ) : (
                                        <Cable sx={{ fontSize: '4rem' }} />
                                    )}
                                </Box>

                                <Typography align="center" marginBottom={1}>
                                    {row.title}
                                </Typography>

                                <Button
                                    href={
                                        row.connector_tags[0].documentation_url
                                    }
                                    target="_blank"
                                    rel="noopener"
                                    endIcon={<OpenInNew />}
                                    sx={{
                                        'backgroundColor': slate[25],
                                        '&:hover': {
                                            backgroundColor: slate[25],
                                        },
                                    }}
                                >
                                    <FormattedMessage id="terms.documentation" />
                                </Button>

                                <Typography
                                    component="div"
                                    variant="caption"
                                    align="center"
                                    marginBottom={5}
                                >
                                    <span
                                        style={{
                                            marginRight: '.5rem',
                                            fontWeight: 'bold',
                                        }}
                                    >
                                        <FormattedMessage id="entityTable.data.lastUpdatedWithColon" />
                                    </span>

                                    <FormattedDate
                                        day="numeric"
                                        month="long"
                                        year="numeric"
                                        value={row.updated_at}
                                    />
                                </Typography>

                                <Stack
                                    direction="row"
                                    spacing={1}
                                    sx={{
                                        alignItems: 'center',
                                        justifyContent: 'flex-start',
                                    }}
                                >
                                    <Button
                                        onClick={() => primaryCtaClick(row)}
                                    >
                                        {row.connector_tags[0].protocol ===
                                        'capture' ? (
                                            <FormattedMessage id="connectorTable.actionsCta.capture" />
                                        ) : (
                                            <FormattedMessage id="connectorTable.actionsCta.materialization" />
                                        )}
                                    </Button>
                                </Stack>
                            </Tile>
                        </Grid>
                    ))
                    .concat(
                        <Grid
                            key="connector-request-tile"
                            item
                            xs={6}
                            md={12 / cardsPerRow}
                        >
                            <Tile>
                                <Box>
                                    <Box sx={imageBackgroundSx}>
                                        <AddBox sx={{ fontSize: '4rem' }} />
                                    </Box>

                                    <Typography align="center" marginBottom={1}>
                                        <FormattedMessage id="connectorTable.data.connectorRequest" />
                                    </Typography>

                                    <Typography
                                        component="p"
                                        variant="caption"
                                        marginBottom={2}
                                        sx={{ px: 1 }}
                                    >
                                        <FormattedMessage id="connectors.main.message2.alt" />
                                    </Typography>
                                </Box>

                                <Button
                                    href={intl.formatMessage({
                                        id: 'connectors.main.message2.docPath',
                                    })}
                                    target="_blank"
                                    rel="noopener"
                                    endIcon={<OpenInNew />}
                                >
                                    <FormattedMessage id="connectorTable.actionsCta.connectorRequest" />
                                </Button>
                            </Tile>
                        </Grid>
                    )
            ) : isValidating || tableState.status === TableStatuses.LOADING ? (
                Array(skeletonTileCount)
                    .fill(
                        <Tile>
                            <Box>
                                <Skeleton
                                    variant="rectangular"
                                    height={125}
                                    sx={{ mb: 2, borderRadius: 3 }}
                                />

                                <Skeleton />

                                <Skeleton />

                                <Skeleton sx={{ mb: 5 }} />
                            </Box>

                            <Skeleton
                                variant="rectangular"
                                height={36}
                                sx={{ borderRadius: 3 }}
                            />
                        </Tile>
                    )
                    .map((skeleton, index) => (
                        <Grid
                            key={`connector-skeleton-${index}`}
                            item
                            xs={6}
                            md={12 / cardsPerRow}
                        >
                            {skeleton}
                        </Grid>
                    ))
            ) : (
                <Grid item sx={{ width: '100%' }}>
                    <Paper
                        elevation={0}
                        sx={{
                            height: 300,
                            borderRadius: 5,
                            display: 'flex',
                            flexDirection: 'column',
                            justifyContent: 'center',
                            background:
                                theme.palette.mode === 'dark'
                                    ? darkGlassBkgColor
                                    : slate[50],
                            padding: 1,
                        }}
                    >
                        <Typography variant="h6" align="center" sx={{ mb: 2 }}>
                            <FormattedMessage
                                id={getEmptyTableHeader(
                                    tableState.status,
                                    intlConfig
                                )}
                            />
                        </Typography>

                        <Typography component="div" align="center">
                            {getEmptyTableMessage(
                                tableState.status,
                                intlConfig
                            )}
                        </Typography>
                    </Paper>
                </Grid>
            )}
        </Grid>
    );
}

export default ConnectorTiles;
