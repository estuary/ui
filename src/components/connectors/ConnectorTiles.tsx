import {
    Grid,
    Paper,
    Typography,
    useMediaQuery,
    useTheme,
} from '@mui/material';
import { getConnectors } from 'api/connectors';
import ConnectorCard from 'components/connectors/card';
import ConnectorToolbar from 'components/connectors/ConnectorToolbar';
import useEntityCreateNavigate from 'components/shared/Entity/hooks/useEntityCreateNavigate';
import { semiTransparentBackground } from 'context/Theme';
import { useSelectNew } from 'hooks/supabase-swr/hooks/useSelect';
import { ConnectorWithTagDetailQuery } from 'hooks/useConnectorWithTagDetail';
import { useEffect, useMemo, useRef, useState } from 'react';
import { FormattedMessage } from 'react-intl';
import { checkErrorMessage, FAILED_TO_FETCH } from 'services/shared';

import {
    EntityWithCreateWorkflow,
    SortDirection,
    TableIntlConfig,
    TableState,
    TableStatuses,
} from 'types';
import { hasLength } from 'utils/misc-utils';
import { getEmptyTableHeader, getEmptyTableMessage } from 'utils/table-utils';
import ConnectorCardDetails from './card/Details';
import ConnectorLogo from './card/Logo';
import ConnectorCardTitle from './card/Title';
import ConnectorRequestTile from './ConnectorRequestTile';
import ConnectorsSkeleton from './Skeleton';

interface ConnectorTilesProps {
    protocolPreset?: EntityWithCreateWorkflow;
    replaceOnNavigate?: boolean;
}

const intlConfig: TableIntlConfig = {
    header: 'connectors.main.message1',
    message: 'connectors.main.message2',
};

function ConnectorTiles({
    protocolPreset,
    replaceOnNavigate,
}: ConnectorTilesProps) {
    const navigateToCreate = useEntityCreateNavigate();
    const isFiltering = useRef(false);

    const theme = useTheme();
    const belowMd = useMediaQuery(theme.breakpoints.down('md'));

    const [protocol, setProtocol] = useState<string | null>(
        protocolPreset ?? null
    );
    const [searchQuery, setSearchQuery] = useState<string | null>(null);
    const [sortDirection, setSortDirection] = useState<SortDirection>('asc');

    const [tableState, setTableState] = useState<TableState>({
        status: TableStatuses.LOADING,
    });

    const query = useMemo(() => {
        return getConnectors(searchQuery, sortDirection, protocol);
    }, [searchQuery, sortDirection, protocol]);

    const {
        data: useSelectResponse,
        isValidating,
        error,
    } = useSelectNew<ConnectorWithTagDetailQuery>(query);

    const selectData = useMemo(
        () => (useSelectResponse ? useSelectResponse.data : []),
        [useSelectResponse]
    );

    const primaryCtaClick = (row: ConnectorWithTagDetailQuery) => {
        navigateToCreate(
            row.connector_tags[0].protocol,
            row.connector_tags[0].connector_id,
            replaceOnNavigate,
            row.connector_tags[0].protocol === 'capture'
        );
    };

    useEffect(() => {
        if (selectData.length > 0) {
            setTableState({ status: TableStatuses.DATA_FETCHED });
        } else if (isFiltering.current) {
            setTableState({ status: TableStatuses.UNMATCHED_FILTER });
        } else if (checkErrorMessage(FAILED_TO_FETCH, error?.message)) {
            setTableState({ status: TableStatuses.NETWORK_FAILED });
        } else {
            setTableState({ status: TableStatuses.NO_EXISTING_DATA });
        }
    }, [selectData, isValidating, error?.message]);

    return (
        <Grid
            container
            spacing={{ xs: 2, md: 3 }}
            columns={{ xs: 4, sm: 4, md: 12, lg: 12, xl: 12 }}
            paddingRight={2}
            margin="auto"
        >
            <Grid item xs={12}>
                <ConnectorToolbar
                    belowMd={belowMd}
                    gridSpacing={2}
                    hideProtocol={!!protocolPreset}
                    setProtocol={setProtocol}
                    setSortDirection={setSortDirection}
                    setSearchQuery={setSearchQuery}
                />
            </Grid>

            {hasLength(selectData) ? (
                selectData
                    .map((row, index) => (
                        <ConnectorCard
                            key={`connector-tile-${index}`}
                            docsUrl={row.connector_tags[0].documentation_url}
                            logo={<ConnectorLogo imageSrc={row.image} />}
                            title={<ConnectorCardTitle title={row.title} />}
                            details={
                                <ConnectorCardDetails
                                    description={row.detail}
                                />
                            }
                            recommended={row.recommended}
                            specType={row.connector_tags[0].protocol}
                            clickHandler={() => primaryCtaClick(row)}
                        />
                    ))
                    .concat(
                        <ConnectorRequestTile key="connector-tile-request" />
                    )
            ) : isValidating || tableState.status === TableStatuses.LOADING ? (
                <ConnectorsSkeleton />
            ) : (
                <Grid item sx={{ width: '100%' }}>
                    <Paper
                        elevation={0}
                        sx={{
                            height: 300,
                            borderRadius: 5,
                            display: 'flex',
                            flexDirection: 'column',
                            justifyContent: 'center',
                            background:
                                semiTransparentBackground[theme.palette.mode],
                            padding: 1,
                        }}
                    >
                        <Typography variant="h6" align="center" sx={{ mb: 2 }}>
                            <FormattedMessage
                                id={getEmptyTableHeader(
                                    tableState.status,
                                    intlConfig
                                )}
                            />
                        </Typography>

                        <Typography component="div" align="center">
                            {getEmptyTableMessage(
                                tableState.status,
                                intlConfig
                            )}
                        </Typography>
                    </Paper>
                </Grid>
            )}
        </Grid>
    );
}

export default ConnectorTiles;
