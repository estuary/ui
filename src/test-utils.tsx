// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render, RenderOptions } from '@testing-library/react';
import { ReactElement, type FC } from 'react';
import { act } from 'react-dom/test-utils';
import { localStorageKey } from './auth';
import AppProviders from './context';

const AllTheProviders: FC = ({ children }) => {
    return <AppProviders>{children}</AppProviders>;
};

const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

const loginAsUser = (userName: string = 'fakeUserName') => {
    act(() => {
        window.localStorage.setItem(localStorageKey, userName);
    });
};

export * from '@testing-library/react';
export { customRender as render, loginAsUser };
