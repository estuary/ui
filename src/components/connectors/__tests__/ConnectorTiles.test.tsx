import { render } from 'test/test-utils';
import { describe, it, expect } from 'vitest';
import ConnectorTiles from '../ConnectorTiles';

describe('Connector Tiles', () => {
    it('should display the connector details', () => {
        const rendered = render(<ConnectorTiles />);
        expect(rendered.baseElement).toMatchSnapshot();
    });
});
