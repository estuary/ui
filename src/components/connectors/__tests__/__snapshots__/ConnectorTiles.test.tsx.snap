// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Connector Tiles > should display the connector details 1`] = `
<body>
  <div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-spacing-md-3 css-1vdl6dg-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-p51zqw-MuiGrid-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular css-sliv39-MuiToolbar-root"
          data-public="true"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1n5bjej-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4 css-cvtzlr-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1juxxoa-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-il7x2h-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="true"
                  for="mui-1"
                  id="mui-1-label"
                >
                  Search connectors
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl MuiInputBase-sizeSmall css-1gcqmk3-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                    id="mui-1"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-14lo706"
                    >
                      <span>
                        Search connectors
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-md-2 css-rvd90y-MuiGrid-root"
            >
              <div
                class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1qqsdnr-MuiAutocomplete-root"
              >
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-il7x2h-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="true"
                    for="mui-3"
                    id="mui-3-label"
                  >
                    Sort Name
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1iez38t-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-expanded="false"
                      aria-invalid="false"
                      autocapitalize="none"
                      autocomplete="off"
                      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
                      id="mui-3"
                      role="combobox"
                      spellcheck="false"
                      type="text"
                      value="A to Z"
                    />
                    <div
                      class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                    >
                      <button
                        aria-label="Open"
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-ytquw7-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                        tabindex="-1"
                        title="Open"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="ArrowDropDownIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M7 10l5 5 5-5z"
                          />
                        </svg>
                      </button>
                    </div>
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-14lo706"
                      >
                        <span>
                          Sort Name
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 MuiGrid-grid-md-2 css-rvd90y-MuiGrid-root"
            >
              <div
                class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1qqsdnr-MuiAutocomplete-root"
              >
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-il7x2h-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="true"
                    for="mui-5"
                    id="mui-5-label"
                  >
                    Protocol
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1iez38t-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-expanded="false"
                      aria-invalid="false"
                      autocapitalize="none"
                      autocomplete="off"
                      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
                      id="mui-5"
                      role="combobox"
                      spellcheck="false"
                      type="text"
                      value="All"
                    />
                    <div
                      class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                    >
                      <button
                        aria-label="Open"
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-ytquw7-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                        tabindex="-1"
                        title="Open"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="ArrowDropDownIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M7 10l5 5 5-5z"
                          />
                        </svg>
                      </button>
                    </div>
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-14lo706"
                      >
                        <span>
                          Protocol
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 MuiGrid-grid-md-4 MuiGrid-grid-lg-2 MuiGrid-grid-xl-2 css-1n6eflt-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-113vizs-MuiPaper-root"
        >
          <div
            class="css-nen11g-MuiStack-root"
            style="height: 100%; justify-content: space-between;"
          >
            <div
              class="MuiBox-root css-0"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-e5s2mh-MuiSkeleton-root"
                style="height: 125px;"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1v30ypl-MuiSkeleton-root"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1v30ypl-MuiSkeleton-root"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1dtiui6-MuiSkeleton-root"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 MuiGrid-grid-md-4 MuiGrid-grid-lg-2 MuiGrid-grid-xl-2 css-1n6eflt-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-113vizs-MuiPaper-root"
        >
          <div
            class="css-nen11g-MuiStack-root"
            style="height: 100%; justify-content: space-between;"
          >
            <div
              class="MuiBox-root css-0"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-e5s2mh-MuiSkeleton-root"
                style="height: 125px;"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1v30ypl-MuiSkeleton-root"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1v30ypl-MuiSkeleton-root"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1dtiui6-MuiSkeleton-root"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 MuiGrid-grid-md-4 MuiGrid-grid-lg-2 MuiGrid-grid-xl-2 css-1n6eflt-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-113vizs-MuiPaper-root"
        >
          <div
            class="css-nen11g-MuiStack-root"
            style="height: 100%; justify-content: space-between;"
          >
            <div
              class="MuiBox-root css-0"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-e5s2mh-MuiSkeleton-root"
                style="height: 125px;"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1v30ypl-MuiSkeleton-root"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1v30ypl-MuiSkeleton-root"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1dtiui6-MuiSkeleton-root"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 MuiGrid-grid-md-4 MuiGrid-grid-lg-2 MuiGrid-grid-xl-2 css-1n6eflt-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-113vizs-MuiPaper-root"
        >
          <div
            class="css-nen11g-MuiStack-root"
            style="height: 100%; justify-content: space-between;"
          >
            <div
              class="MuiBox-root css-0"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-e5s2mh-MuiSkeleton-root"
                style="height: 125px;"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1v30ypl-MuiSkeleton-root"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1v30ypl-MuiSkeleton-root"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1dtiui6-MuiSkeleton-root"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 MuiGrid-grid-md-4 MuiGrid-grid-lg-2 MuiGrid-grid-xl-2 css-1n6eflt-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-113vizs-MuiPaper-root"
        >
          <div
            class="css-nen11g-MuiStack-root"
            style="height: 100%; justify-content: space-between;"
          >
            <div
              class="MuiBox-root css-0"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-e5s2mh-MuiSkeleton-root"
                style="height: 125px;"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1v30ypl-MuiSkeleton-root"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1v30ypl-MuiSkeleton-root"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1dtiui6-MuiSkeleton-root"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 MuiGrid-grid-md-4 MuiGrid-grid-lg-2 MuiGrid-grid-xl-2 css-1n6eflt-MuiGrid-root"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-113vizs-MuiPaper-root"
        >
          <div
            class="css-nen11g-MuiStack-root"
            style="height: 100%; justify-content: space-between;"
          >
            <div
              class="MuiBox-root css-0"
            >
              <span
                class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-e5s2mh-MuiSkeleton-root"
                style="height: 125px;"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1v30ypl-MuiSkeleton-root"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1v30ypl-MuiSkeleton-root"
              />
              <span
                class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1dtiui6-MuiSkeleton-root"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
